### PR TITLE
Handle htlc loss when looking for final HTLC resolution

### PR DESCRIPTION
**Pull request by Torakushi**
When looking for a HTLC final resolution, it's possible that LND lost the HTLC and returns a `HtlcEvent_FinalHtlcEvent` with settled == false.

In this case, we will mark the HTLC and the invoice as 'FAILED' and we will notify the user using the `WaitForInvoiceFinalEvent` (Formerly WaitForInvoiceSettled)

To do that we:

- added more generic columns to track the invoice resolution
- added the invoice status in the `WaitForInvoiceFinalEventResponse`

### CODE REVIEW

File: persistence/pg.go [hunk 1]
```diff
@@ -260,68 +262,97 @@ func (p *PostgresPersister) MarkHtlcSettled(ctx context.Context,
...
 
-		now := time.Now().UTC()
-
-		htlc := dbHtlc{
-			Node:   key.Node,
-			ChanID: key.ChanID,
-			HtlcID: key.HtlcID,
+		// It is possible that we received previously a HTLC that has
+		// failed for this invoice. In consequence, the invoice should
+		// have the FAILED status. However, a HTLC could have been settled
```
Comment (Torakushi) [inline persistence/pg.go]: I am not sure if this is a real case, I could not find anything explicit in LND.

File: db.go [hunk 2]
```diff
@@ -43,11 +43,6 @@ var (
 type Invoice struct {
 	types.InvoiceCreationData
 
-	Settled bool
-
-	// SettleDate is the exact time the invoice was settled.
-	SettleDate time.Time
```
Comment (joostjager) [inline db.go]: Can move this into a separate PR on master to get it out of the way

File: lnmuxrpc/lnmux.proto [hunk 3]
```diff
@@ -94,7 +94,7 @@ message Invoice {
     bytes preimage = 2;
     uint64 amount_msat = 3;
     bool settled = 4;
-    uint64 settled_requested_at = 5;
+    uint64 settle_requested_at = 5;
```
Comment (joostjager) [inline lnmuxrpc/lnmux.proto:106]: Does the deposit connector need to be updated too?

File: cmd/lnmuxd/server.go [hunk 4]
```diff
@@ -336,9 +336,25 @@ func dbInvoiceToProto(invoice *persistence.Invoice) *lnmuxrpc.Invoice {
...
+	var status lnmuxrpc.InvoiceStatus
+	switch invoiceStatus {
+	case types.StatusSettleRequested:
+		status = lnmuxrpc.InvoiceStatus_SETTLE_REQUESTED
+	case types.StatusSettled:
+		status = lnmuxrpc.InvoiceStatus_SETTLED
+	case types.StatusFailed:
+		status = lnmuxrpc.InvoiceStatus_FAILED
+	default:
+		status = lnmuxrpc.InvoiceStatus_UNKNOWN
```
Comment (joostjager) [inline cmd/lnmuxd/server.go]: Is it happening, UNKNOWN?

File: persistence/migrations/4_add_status.up.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+ALTER TABLE lnmux.htlcs ADD "status" VARCHAR;
+ALTER TABLE lnmux.htlcs ADD "resolution_at" TIMESTAMPTZ;
+
+
+ALTER TABLE lnmux.invoices DROP COLUMN "settled";
+ALTER TABLE lnmux.invoices DROP COLUMN "settled_at";
+
+ALTER TABLE lnmux.htlcs DROP COLUMN "settled";
+ALTER TABLE lnmux.htlcs DROP COLUMN "settled_at";
```
Comment (joostjager) [inline persistence/migrations/4_add_status.up.sql:34]: Data migration from old to new column?

File: persistence/pg.go [hunk 6]
```diff
@@ -21,8 +21,9 @@ type Invoice struct {
 
 	SequenceNum       uint64
 	SettleRequestedAt time.Time
-	SettledAt         time.Time
-	Settled           bool
+	ResolutionAt      time.Time
+	Status            types.Status
+	ReachFinalStatus  bool
```
Comment (joostjager) [inline persistence/pg.go]: Maybe this can be a method `IsFinal()` on `types.Status`? It looks redundant to store in the database.

File: cmd/lnmuxd/server.go [hunk 7]
```diff
@@ -155,15 +155,15 @@ func (s *server) SubscribeInvoiceAccepted(req *lnmuxrpc.SubscribeInvoiceAccepted
...
+	req *lnmuxrpc.WaitForInvoiceFinalEventRequest) (
+	*lnmuxrpc.WaitForInvoiceFinalEventResponse, error) {
 
 	hash, err := lntypes.MakeHash(req.Hash)
 	if err != nil {
 		return nil, err
 	}
 
-	err = s.settledHandler.WaitForInvoiceSettled(ctx, hash)
+	invoiceStatus, err := s.settledHandler.WaitForInvoiceSettled(ctx, hash)
```
Comment (joostjager) [inline cmd/lnmuxd/server.go]: It's not renamed as promised by the commit message?

File: cmd/lnmuxd/server.go [hunk 8]
```diff
@@ -174,7 +174,9 @@ func (s *server) WaitForInvoiceSettled(ctx context.Context,
 		return nil, status.Errorf(codes.Internal, err.Error())
 	}
 
-	return &lnmuxrpc.WaitForInvoiceSettledResponse{}, nil
+	return &lnmuxrpc.WaitForInvoiceFinalEventResponse{
+		InvoiceStatus: statusToProto(invoiceStatus),
```
Comment (joostjager) [inline cmd/lnmuxd/server.go]: Is this strict enough when only `SETTLED` and `FAILED` are returned? Alternatively it could be a boolean. But maybe the consistency with other calls is a plus too.

File: cmd/lnmuxd/server.go [hunk 9]
```diff
@@ -155,15 +155,15 @@ func (s *server) SubscribeInvoiceAccepted(req *lnmuxrpc.SubscribeInvoiceAccepted
 }
 
 func (s *server) WaitForInvoiceSettled(ctx context.Context,
```
Comment (joostjager) [inline cmd/lnmuxd/server.go]: WaitForInvoiceFinalEvent?

File: settled_handler.go [hunk 10]
```diff
@@ -64,22 +65,27 @@ func (p *SettledHandler) WaitForInvoiceSettled(ctx context.Context,
 	// Check database to see if invoice was already settled.
 	invoice, _, err := p.persister.Get(ctx, hash)
 	if err != nil {
-		return err
+		return types.StatusUnknown, err
```
Comment (joostjager) [inline settled_handler.go]: For enums that don't matter, I usually return `0, err`. To make it clear that it really does not matter what the primary return value is.

File: node_settled_handler.go [hunk 11]
```diff
@@ -115,25 +110,60 @@ func (p *NodeSettledHandler) subscribeEvents(ctx context.Context) error {
...
+
+		if err := p.handleFinalHtlc(ctx, key, failedResolution); err != nil {
 			return err
 		}
 	}
 }
 
 func (p *NodeSettledHandler) handleFinalHtlc(ctx context.Context,
-	key types.CircuitKey) error {
+	key types.CircuitKey, failedResolution bool) error {
```
Comment (joostjager) [inline node_settled_handler.go]: Mostly `settled bool` is used, but this is `failed bool`. Maybe better to keep it consistent and always use `settled`.

File: node_settled_handler.go [hunk 12]
```diff
@@ -115,25 +110,60 @@ func (p *NodeSettledHandler) subscribeEvents(ctx context.Context) error {
...
+		}
+
+		status = types.StatusFailed
+	} else {
+		invoiceHash, err = p.handleFinalSettledHtlc(ctx, key)
+		if err != nil {
+			return err
+		}
+
+		status = types.StatusSettled
```
Comment (joostjager) [inline node_settled_handler.go]: If this is just for logging, maybe you can make a logger at the top of this function using `logger.With`, and then log inside the if blocks `Infow("final final received")` and `Infow("final settle received")`?

File: node_settled_handler.go [hunk 13]
```diff
@@ -143,24 +173,43 @@ func (p *NodeSettledHandler) handleFinalHtlc(ctx context.Context,
 
 	settledHash, err := p.persister.MarkHtlcSettled(ctx, htlcKey)
 	switch {
-	case err == persistence.ErrHtlcAlreadySettled:
-		return nil
+	case err == persistence.ErrHtlcAlreadyInStatus:
```
Comment (joostjager) [inline node_settled_handler.go]: I see you want to reuse the err message, but this name isn't optimally clear I think. Perhaps `ErrHtlcAlreadyFinal`?

File: node_settled_handler.go [hunk 14]
```diff
@@ -143,24 +173,43 @@ func (p *NodeSettledHandler) handleFinalHtlc(ctx context.Context,
...
-	if settledHash != nil {
-		p.settledCallback(*settledHash)
+	htlcKey := types.HtlcKey{
+		ChanID: key.ChanID,
+		HtlcID: key.HtlcID,
+		Node:   p.lnd.PubKey(),
 	}
 
-	return nil
+	failedHash, err := p.persister.MarkHtlcFailed(ctx, htlcKey)
```
Comment (joostjager) [inline node_settled_handler.go]: Instead of this duplication, maybe you can do a `persister.MarkHtlcFinal(ctx, htlckey, settled)`?

File: persistence/pg.go [hunk 15]
```diff
@@ -331,15 +362,66 @@ func (p *PostgresPersister) MarkHtlcSettled(ctx context.Context,
...
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	return settledHash, nil
+	return invoiceHash, nil
+}
+
+func (p *PostgresPersister) selectInvoiceForUpdate(ctx context.Context,
```
Comment (joostjager) [inline persistence/pg.go:335]: Refactor can go to a separate commit?

File: persistence/pg.go [hunk 16]
```diff
@@ -260,68 +262,97 @@ func (p *PostgresPersister) MarkHtlcSettled(ctx context.Context,
...
-		htlc := dbHtlc{
-			Node:   key.Node,
-			ChanID: key.ChanID,
-			HtlcID: key.HtlcID,
+		// It is possible that we received previously a HTLC that has
+		// failed for this invoice. In consequence, the invoice should
+		// have the FAILED status. However, a HTLC could have been settled
+		// afterward.
+		if invoice.Status == types.StatusFailed {
+			return nil
```
Comment (joostjager) [inline persistence/pg.go]: By returning `nil` another notification is prevented?

File: persistence/pg.go [hunk 17]
```diff
@@ -260,68 +262,97 @@ func (p *PostgresPersister) MarkHtlcSettled(ctx context.Context,
...
+			Where("hash=?", invoice.Hash).
 			Set("status=?", types.StatusSettled).
 			Set("resolution_at=?", now).
 			Update() // nolint:contextcheck
 		if err != nil {
 			return err
 		}
 		if result.RowsAffected() == 0 {
-			return ErrHtlcAlreadySettled
+			return types.ErrInvoiceNotFound
```
Comment (joostjager) [inline persistence/pg.go]: I thought at this point the invoice should always be there?

File: settled_handler.go [hunk 18]
```diff
@@ -64,22 +65,27 @@ func (p *SettledHandler) WaitForInvoiceSettled(ctx context.Context,
...
+		return invoice.Status, nil
 	}
 
 	// Not settled yet. Wait for the event.
 	select {
 	case <-waitChan:
 		logger.Debugw("Wait for invoice settled completed via wait channel")
 
-		return nil
+		invoice, _, err := p.persister.Get(ctx, hash)
```
Comment (joostjager) [inline settled_handler.go]: Is this re-fetch necessary or can the final outcome be passed over the `waitChan`?

File: persistence/pg.go (hunk 1 above)
Comment (joostjager) [reply persistence/pg.go]: I think this is real.

File: lnmuxrpc/lnmux.proto (hunk 3 above)
Comment (Torakushi) [reply lnmuxrpc/lnmux.proto:106]: No, this `message` is only use for the ListInvoices rpc call

File: persistence/migrations/4_add_status.up.sql (hunk 5 above)
Comment (Torakushi) [reply persistence/migrations/4_add_status.up.sql:34]: What do you mean exactly?

I removed settled and settled_at in both invoices and htlcs tables and added
status and resolution_at in both tables?

the migration down is done in 4_add_status.down.sql

File: persistence/pg.go (hunk 6 above)
Comment (Torakushi) [reply persistence/pg.go]: `ReachFinalStatus` is never stored in the DB (we use the `dbInvoice` structure to store in DB)
`ReachFinalStatus` is only a field to know when the invoice reaches its final status.

But I do agree, having IsFinal() is better than a new field!

File: cmd/lnmuxd/server.go (hunk 9 above)
Comment (Torakushi) [reply cmd/lnmuxd/server.go]: oh.... nice catch

File: cmd/lnmuxd/server.go (hunk 7 above)
Comment (Torakushi) [reply cmd/lnmuxd/server.go]: I didn't want to rename it in the `settledHandler` handler because (firstly), I thought that it is normal for a settledHandler to wait for settled 😄 

And, I also thought that, the failed status is a specific case occurring when we wait for the settled signal.

But I would be quite happy to rename it here too!

File: cmd/lnmuxd/server.go (hunk 8 above)
Comment (Torakushi) [reply cmd/lnmuxd/server.go]: Hmm you are right. In this case, we are only waiting for SETTLED or FAILED,

I will restrict `statusToProto` cases.

Personaly, I would prefer keeping this enum (for consistency like you said). But I can totally make it a boolean!

File: cmd/lnmuxd/server.go (hunk 4 above)
Comment (Torakushi) [reply cmd/lnmuxd/server.go]: linked to this comment: https://github.com/bottlepay/lnmux/pull/131/files#r1144875714

I will restrict the cases here.

Unknown is only return by the settledHandler in case of error

File: persistence/pg.go (hunk 16 above)
Comment (Torakushi) [reply persistence/pg.go]: By returning nil, we avoid to mark an invoice as settled if it has been already marked as failed.

Do you prefer to return an error and add a log in the caller?

File: persistence/pg.go (hunk 17 above)
Comment (Torakushi) [reply persistence/pg.go]: The github diff is misleading

`return types.ErrInvoiceNotFound` is already present in settle_final branch
for the invoice update:

https://github.com/bottlepay/lnmux/pull/106/files#diff-fa1f5d8279c623bad5b14b09e9cabf2c5121065152f2aaab652bb71a85fab6b7R329

But I can change it if you prefer?

File: settled_handler.go (hunk 18 above)
Comment (Torakushi) [reply settled_handler.go]: First, I didn;t want to because It will need a bit of refactoring
(change 	`settledCallback func(lntypes.Hash)` in node_settle_handler.go, mux.go,...

But I can if you think that the re-fetch is really annoying!

File: lnmuxrpc/lnmux.proto (hunk 3 above)
Comment (joostjager) [reply lnmuxrpc/lnmux.proto:106]: Np then :+1:

File: cmd/lnmuxd/server.go (hunk 8 above)
Comment (joostjager) [reply cmd/lnmuxd/server.go]: On second thought, I think it's fine to keep the enum.

File: cmd/lnmuxd/server.go (hunk 4 above)
Comment (joostjager) [reply cmd/lnmuxd/server.go]: Ah, so that will never be mapped to an actual proto enum. I think you can remove it then?

File: persistence/migrations/4_add_status.up.sql (hunk 5 above)
Comment (joostjager) [reply persistence/migrations/4_add_status.up.sql:34]: I mean migrating the data that is already present in those tables.

File: cmd/lnmuxd/server.go (hunk 7 above)
Comment (joostjager) [reply cmd/lnmuxd/server.go]: I mean `WaitForInvoiceSettled` :)

File: persistence/pg.go (hunk 15 above)
Comment (joostjager) [reply persistence/pg.go:335]: Anything refactor can be pulled out of the PR and merged onto master, if it makes sense.

File: persistence/pg.go (hunk 16 above)
Comment (joostjager) [reply persistence/pg.go]: Ah no, it's ok. I was just checking to see what the effect of `nil` is here. But tracing through the code, I see that ultimately any rpc waiters for settled aren't notified. And this seems good, because the final event already happened earlier.

File: persistence/pg.go (hunk 17 above)
Comment (joostjager) [reply persistence/pg.go]: Ah ok. And what needs to happen to end up here? The invoice is in the database, that has been checked in `selectInvoiceForUpdate`. If it is never supposed to happen, I'd return a different error than `ErrInvoiceNotFound`.

File: settled_handler.go (hunk 18 above)
Comment (joostjager) [reply settled_handler.go]: I'd just do the refactor. I think it also helps with understanding the flow.

File: settled_handler.go (hunk 10 above)
Comment (Torakushi) [reply settled_handler.go]: Never thought of that, nice one, will go for `"", err`

File: persistence/pg.go (hunk 17 above)
Comment (Torakushi) [reply persistence/pg.go]: We will be here if the invoice has state has changed in the meantime (not SETTLE_REQUESTED anymore). So we can't find the invoice with the given state

Should never happen (because we did a selectForUpdate).

File: cmd/lnmuxd/server.go (hunk 8 above)
Comment (Torakushi) [reply cmd/lnmuxd/server.go]: On a second thought as well 😄 : We need to keep `SETTLE_REQUESTED` in the statusToProto because of the `ListInvoices` RPC call

File: node_settled_handler.go (hunk 12 above)
Comment (Torakushi) [reply node_settled_handler.go]: now that the wait channel will return the invoice status, it's not only for logging

File: node_settled_handler.go [hunk 19]
```diff
@@ -29,7 +29,7 @@ type NodeSettledHandler struct {
 
 	newWatchChan chan types.CircuitKey
 
-	settledCallback func(lntypes.Hash)
+	settledCallback func(lntypes.Hash, types.Status)
```
Comment (Torakushi) [inline node_settled_handler.go]: should we rename `settledCallback` as well? IMO, it's not needed but... it's me 😄

File: persistence/pg.go [hunk 20]
```diff
@@ -355,6 +351,40 @@ func (p *PostgresPersister) settleHTLC(ctx context.Context,
 	return nil
 }
 
+func (p *PostgresPersister) changeInvoiceStatus(ctx context.Context,
```
Comment (Torakushi) [inline persistence/pg.go]: should it be in a separate commit? (ideally it should have been in the refactoring PR..)

File: lnmuxrpc/lnmux.proto [hunk 21]
```diff
@@ -89,12 +89,19 @@ message ListInvoicesResponse {
...
+
 message Invoice {
     bytes hash = 1;
     bytes preimage = 2;
     uint64 amount_msat = 3;
-    bool settled = 4;
+    InvoiceStatus status = 4;
     uint64 settle_requested_at = 5;
-    uint64 settled_at = 6;
+    uint64 resolution_at = 6;
```
Comment (joostjager) [inline lnmuxrpc/lnmux.proto]: I have some slight doubts about this name, but also not sure about something better. Maybe `finalized_at`, but that's also not ideal.

`final_at`, `completed_at` maybe?

File: persistence/migrations/4_add_status.up.sql (hunk 5 above)
Comment (joostjager) [reply persistence/migrations/4_add_status.up.sql:34]: Looks good

File: persistence/pg.go (hunk 20 above)
Comment (joostjager) [reply persistence/pg.go]: I think it's better in a separate commit indeed.

File: persistence/migrations/4_add_status.up.sql [hunk 22]
```diff
@@ -0,0 +1,23 @@
+-- Migrate invoices table
+ALTER TABLE lnmux.invoices ADD "status" VARCHAR;
```
Comment (joostjager) [inline persistence/migrations/4_add_status.up.sql]: I think you can make an enum in postgres.

File: cmd/lnmuxd/server.go [hunk 23]
```diff
@@ -154,16 +154,16 @@ func (s *server) SubscribeInvoiceAccepted(req *lnmuxrpc.SubscribeInvoiceAccepted
...
+	req *lnmuxrpc.WaitForInvoiceFinalEventRequest) (
+	*lnmuxrpc.WaitForInvoiceFinalEventResponse, error) {
 
 	hash, err := lntypes.MakeHash(req.Hash)
 	if err != nil {
 		return nil, err
 	}
 
-	err = s.settledHandler.WaitForInvoiceSettled(ctx, hash)
+	invoiceStatus, err := s.settledHandler.WaitForInvoiceFinalEvent(ctx, hash)
```
Comment (joostjager) [inline cmd/lnmuxd/server.go]: `WaitForInvoiceFinalStatus`, to remain consistent with the variable name/status enum?

File: mux.go [hunk 24]
```diff
@@ -36,7 +36,7 @@ type Mux struct {
 type MuxConfig struct {
 	KeyRing         keychain.SecretKeyRing
 	ActiveNetParams *chaincfg.Params
-	SettledCallback func(lntypes.Hash)
+	SettledCallback func(lntypes.Hash, types.Status)
```
Comment (joostjager) [inline mux.go]: Maybe make it `types.InvoiceStatus`. Otherwise confusion will happen with the next status type added.

Comment (joostjager) [reply mux.go]: `HtlcStatus` can be an alias, but it is still nice to have a separation between the types to prevent potential bugs.

File: node_settled_handler.go [hunk 25]
```diff
@@ -115,51 +110,61 @@ func (p *NodeSettledHandler) subscribeEvents(ctx context.Context) error {
...
+		"chanID", key.ChanID,
+		"htlcID", key.HtlcID,
+		"hash", invoiceHash,
+		"status", status,
+	)
 
-	if settledHash != nil {
-		p.settledCallback(*settledHash, types.StatusSettled)
+	if invoiceHash != nil {
+		p.settledCallback(*invoiceHash, status)
```
Comment (joostjager) [inline node_settled_handler.go]: Yes, so here I wonder if it isn't better to use a boolean, to make it more strict?

File: persistence/pg.go [hunk 26]
```diff
@@ -249,53 +251,77 @@ func (p *PostgresPersister) getHtlcHash(ctx context.Context, tx *pg.Tx,
...
 		if err != nil {
 			return err
 		}
-		if !allHtlcsSettled {
+
+		// It is possible that we received previously a HTLC that has
+		// failed for this invoice. In consequence, the invoice should
+		// have the FAILED status. However, a HTLC could have reached
+		// its final status afterward.
+		if invoice.Status == types.StatusFailed {
```
Comment (joostjager) [inline persistence/pg.go]: Still think it is worth adding to the comment here why we don't return an invoice hash for a final invoice.

Comment (joostjager) [inline persistence/pg.go]: Maybe also catch `StatusSettled` here explicitly and return an error? Something must be in an inconsistent state if this happens right?

File: persistence/pg.go (hunk 17 above)
Comment (joostjager) [reply persistence/pg.go]: So then it's better to return a different error? To not hide a sitaution that is never supposed to happen?

File: node_settled_handler.go (hunk 19 above)
Comment (joostjager) [reply node_settled_handler.go]: For consistency yes, `finalCallback` or `completedCallback`?

File: persistence/pg.go (hunk 26 above)
Comment (Torakushi) [reply persistence/pg.go]: In which case this will happen?

I mean, if invoice has StatusSettled state, it means that all HTLC were settled and MarkHtlcFinal will never be called again for this invoice

Comment (joostjager) [reply persistence/pg.go]: Yes, as a sanity check I mean

File: lnmuxrpc/lnmux.proto (hunk 21 above)
Comment (Torakushi) [reply lnmuxrpc/lnmux.proto]: Will work my naming skill

File: persistence/migrations/4_add_status.up.sql (hunk 22 above)
Comment (Torakushi) [reply persistence/migrations/4_add_status.up.sql]: Fair, I was thinking about this.

In that case, we may want to create 2 different enums, one for htlc status and the other one for invoice status
(to be consistent with https://github.com/bottlepay/lnmux/pull/131/commits/49a926f62e7eea351e309e97dace35ab011199bf#r1146231182)

File: lnmuxrpc/lnmux.proto (hunk 21 above)
Comment (joostjager) [reply lnmuxrpc/lnmux.proto]: Just ask your AI assistant to come up with options 😆

File: persistence/migrations/4_add_status.up.sql (hunk 22 above)
Comment (joostjager) [reply persistence/migrations/4_add_status.up.sql]: Yes, I'd opt for separate types.

File: mux.go (hunk 24 above)
Comment (Torakushi) [reply mux.go]: Yeah I do agree on the separation!

To be consistent with what we did in the DB, I will create 2 distinct types (not aliasing otherwise it can be confusing to have a `HTLCStatus` with  type `InvoiceStatus`)

It will ease things if one day we want to change/add a status for either invoice or htlc

File: node_settled_handler.go (hunk 25 above)
Comment (Torakushi) [reply node_settled_handler.go]: After a second thought, maybe it makes more sense here indeed.
More strict, and it makes sense that `settleHandler.InvoiceSettled` sends a bool in the channel instead of a status.

Then the server, depending on the bool, will return the right status (I really prefer to send the status (and not the bool) through the RPC call, it will be confusing the the user otherwise)

File: node_settled_handler.go [hunk 27]
```diff
@@ -115,51 +110,61 @@ func (p *NodeSettledHandler) subscribeEvents(ctx context.Context) error {
...
 			return err
 		}
 	}
 }
 
 func (p *NodeSettledHandler) handleFinalHtlc(ctx context.Context,
-	key types.CircuitKey) error {
+	key types.CircuitKey, settled bool) error {
+
+	var status types.HtlcStatus
```
Comment (Torakushi) [inline node_settled_handler.go]: It's only for logging indeed.

If it's annoying, I can either put the logs directly in persistence (maybe not as consistent), or remove some info from the logs (the status)

File: persistence/pg.go [hunk 28]
```diff
@@ -249,53 +263,87 @@ func (p *PostgresPersister) getHtlcHash(ctx context.Context, tx *pg.Tx,
...
+			return ErrHtlcReceivedButInvoiceAlreadySettled
 		}
 
-		// The HTLC is settled, now we check if all other HTLCs
-		// are settled as well.
-		allHtlcsSettled, err := p.allHtlcsAreSettled(ctx,
-			tx, invoice.Hash)
+		now := time.Now().UTC()
+
+		err = p.changeHTLCStatus(ctx, tx, key, htlcFinalStatus, now)
```
Comment (joostjager) [inline persistence/pg.go]: nit: change sounds like you can change it to anything. Maybe `finalizeHTLC(settled)/finalizeInvoice(settled)`?

File: node_settled_handler.go [hunk 29]
```diff
@@ -115,51 +110,61 @@ func (p *NodeSettledHandler) subscribeEvents(ctx context.Context) error {
...
+		"chanID", key.ChanID,
+		"htlcID", key.HtlcID,
+		"hash", invoiceHash,
+		"status", status,
+	)
 
-	if settledHash != nil {
-		p.finalCallback(*settledHash, true)
+	if invoiceHash != nil {
+		p.finalCallback(*invoiceHash, settled)
```
Comment (joostjager) [inline node_settled_handler.go:157]: Do you want to add test coverage for the invoice failed case?

**Review by joostjager — COMMENTED** [review]
Looks close, lots of improvements made!

File: persistence/pg.go (hunk 17 above)
Comment (joostjager) [reply persistence/pg.go]: I think this has been addressed now right? Can't recall the full context of this comment with the new push.

File: node_settled_handler.go [hunk 30]
```diff
@@ -115,51 +110,61 @@ func (p *NodeSettledHandler) subscribeEvents(ctx context.Context) error {
...
-	// If the htlc is not found, the final resolution was for an htlc that isn't
-	// managed by lnmux.
+	// If the htlc is not found, the final resolution was for an htlc that
+	// isn't managed by lnmux.
 	case err == types.ErrHtlcNotFound:
 		return nil
 
 	case err != nil:
-		return fmt.Errorf("unable to mark htlc %v settled: %w", key, err)
+		return fmt.Errorf("unable to mark htlc %v as %s : %w",
```
Comment (joostjager) [inline node_settled_handler.go]: Remove space before colon

File: node_settled_handler.go (hunk 27 above)
Comment (joostjager) [reply node_settled_handler.go]: I think you can just log the settled boolean which you already have instead of mapping to status just for logging/err?

File: node_settled_handler.go (hunk 29 above)
Comment (Torakushi) [reply node_settled_handler.go:157]: very very weird

I did the tests:

https://github.com/bottlepay/lnmux/commit/2abdd53414e07d4ca53f850c91a24b2daab6b41d#diff-b39fc442f543ad6df386d0772fdbafafee01095292545626a39251403bec6b5cR458

But for some reason, when doing a rebase, I removed them...

Comment (joostjager) [reply node_settled_handler.go:157]: Oh oh.... maybe something else changed. You can diff to some revision from `git reflog`

Comment (Torakushi) [reply node_settled_handler.go:157]: I triple checked, only this test was "lost"!

I just added the test again

File: lnmuxrpc/lnmux.proto [hunk 31]
```diff
@@ -89,12 +89,19 @@ message ListInvoicesResponse {
     uint64 last_sequence_number = 3;
 }
 
+enum InvoiceStatus {
+    UNKNOWN = 0;
+    FAILED = 1;
+    SETTLE_REQUESTED=2;
```
Comment (jjager-nydig) [inline lnmuxrpc/lnmux.proto]: ultra-nit: spaces around equal sign

File: persistence/migrations/4_add_status.up.sql [hunk 32]
```diff
@@ -0,0 +1,34 @@
+CREATE TYPE invoice_status as ENUM (
+    'SETTLE_REQUESTED',
+    'SETTLED',
+    'FAILED'
+);
+
+-- Migrate invoices table
+ALTER TABLE lnmux.invoices ADD "status" invoice_status;
```
Comment (jjager-nydig) [inline persistence/migrations/4_add_status.up.sql]: NOT NULL here and for htlcs?

File: persistence/migrations/4_add_status.up.sql [hunk 33]
```diff
@@ -0,0 +1,34 @@
...
+ALTER TABLE lnmux.invoices ADD "finalized_at" TIMESTAMPTZ;
+
+UPDATE lnmux.invoices SET
+        status = CASE WHEN settled = FALSE THEN 'SETTLE_REQUESTED'::invoice_status ELSE 'SETTLED'::invoice_status END,
+        finalized_at = CASE WHEN settled = TRUE THEN settled_at ELSE NULL END;
+
+ALTER TABLE lnmux.invoices DROP COLUMN "settled";
+ALTER TABLE lnmux.invoices DROP COLUMN "settled_at";
```
Comment (jjager-nydig) [inline persistence/migrations/4_add_status.up.sql:16]: I think that postgres also has a RENAME COLUMN statement. But this works too.

File: persistence/pg.go [hunk 34]
```diff
@@ -378,7 +412,7 @@ func (p *PostgresPersister) finalizeInvoice(ctx context.Context,
 		return err
 	}
 	if result.RowsAffected() == 0 {
-		return types.ErrInvoiceNotFound
+		return ErrInvoiceAlreadyFinal
```
Comment (jjager-nydig) [inline persistence/pg.go]: Hmm, looking at this function in isolation, I am not sure about this change.

Maybe the update needs a where clause checking for `SettleRequested` too.

I suppose you can't distinguish between a non-existing invoice and an invoice that is already final in one query.

**Review by joostjager — COMMENTED** [review]
Unintentionally reviewed as @jjager-nydig

File: persistence/migrations/4_add_status.up.sql (hunk 33 above)
Comment (Torakushi) [reply persistence/migrations/4_add_status.up.sql:16]: Indeed, I didn't know it exists. It could be convenient for the migrate_up, but for the migrate_down, we still need to use the DROP (or a more complex logic) because for invoice with 'FAILED' status, `finalized_at` will be not NULL but we don't want `settled_at` to be not null (I think) 

I think I will keep the drop for now (if really no objections)

File: persistence/pg.go (hunk 34 above)
Comment (Torakushi) [reply persistence/pg.go]: when calling `finalizeInvoice` we always do a SELECT FOR UPDATE, in the same transaction.

So, the only case where `result.RowsAffected() == 0` are:

- invoice already in invoiceFinalStatus state
- invoice was removed from the DB manually in the meantime

However, I think that it's important to add a Where clause (`WHERE status = SETTLE_REQUESTED`). Because, IMO (even if it should never happen), `finalizeinvoice` should not change the status of an invoice that has already reached it's final status.
(currently if invoice is marked as FAILED, and we call `finalizeinvoice(SETTLED)`, it will change the invoice final status)

I've added this where statement.
Did I miss the point? (if the goal is to really know if the invoice wasn't deleted in the meantime, we can re-do a SELECT statement. But we still have no guarantee, we only narrow the gap)

Comment (joostjager) [reply persistence/pg.go]: Yes, I know that those outcomes are the only ones. But I mean that just looking at this function in isolation, the context with select-for-update isn't visible.

Maybe it is enough to just rename the error to something that is always true for this function. Which is inevitably something more generic such as `ErrFinalizationFailed`. But then at least you can read this function without assumptions about the external environment.

Comment (Torakushi) [reply persistence/pg.go]: ahhh, didn't understand that, will do!

**Review by joostjager — APPROVED** [review]

**Comment by joostjager** [comment]
One more question, did you manually test the functionality with `lnd` nodes? This can be done by bringing down nodes at the right time, so that the peer is going to claw back the htlc and the invoice is going to end up in the failed state in the lnmux db.

**Comment by Torakushi** [comment]
Yes! tested an checked
